### ISSUE TITLE
Implement Scxhulze STV proportional method

### ISSUE THREAD

**Issue by julien-boudry** [julien-boudry/Condorcet#64]
https://en.wikipedia.org/wiki/Schulze_STV

**Comment by LiamM32** [comment]
Yes! Do it! That would be great.

**Comment by julien-boudry** [comment]
Looking for complete step by steps documentation. And test vectors.

**Comment by LiamM32** [comment]
It would be nice if there was an option for a semi-proportional form of Schulze STV, with an option to set the level of proportionality desired. But you would have to figure-out how that would work mathematically.

**Comment by julien-boudry** [comment]
If a method is correctly specified, published, well recognized. So... why not ;)

The first step should be to implement te existing Schulze STV method.

**Comment by LiamM32** [comment]
I think I've figured out how you can make Schulze STV semi-proportional, if you have the regular version. Before determining the winner, for every pair of candidates, delete an equal number of votes for each candidate; a percentage of the votes given to the less-supported candidate. Therefore the margins would remain the same (as defined in the single-winner version). But it would increase the ratio of support.

I haven't read it yet, but this appears to be a paper explaining the process: [](https://aso.icann.org/wp-content/uploads/2019/02/schulze3.pdf)

**Comment by LiamM32** [comment]
Here's a paper that looks simpler than the one above. It includes the Schulze Method, Schulze Proportional Ranking, & Schulze STV. The latter two would be great additions to Condorcet, but I struggle to understand how the process is described in these papers.

**Comment by julien-boudry** [comment]
Yes, not so easy at all. I'm bad for pure math explanation. I'm strong to understand the logic, step by step, and its implications for implementation. But not abstract formula.

Methods that can't be explained easily to non-expert humans are strange. Event us, a kind of specialist, has difficulty understanding it.

**Comment by LiamM32** [comment]
Yeah. Given how widely known Schulze STV is in voting theory circles, it's surprising how little has been written to explain the process. I'm unfortunately not familiar with all those mathematical symbols.

But there is another software project that claims to have implemented Schulze STV; Python-Vote. Here is the file: https://github.com/the-maldridge/python-vote-core/blob/master/py3votecore/schulze_stv.py

**Comment by LiamM32** [comment]
I cannot guarantee that this is the same comparison method used in the paper, but I think it's likely that it is. In the absence of understanding of the language used in those papers, I may have recreated the technique in my head for pairwise comparisons.

I know that Schulze STV only determines the direct paths between sets that differ by only one candidate. So imagine this is a 3-winner Schulze STV election. Our quota would be a quarter of voters. We are comparing a set containing Loreen, Bob, & Katherine to a set containing Loreen, Bob, & Pierre.

First you must determine for each candidate in either set, how many voters prefer that candidate to all of the other 3. If Bob & Loreen each have more than a quota of voters who prefer them to the others, you would transfer a portion of their votes to either Katherine or Pierre. These combined votes for either Katherine or Pierre are used to compare the two sets.

If you don't follow, I can give a pseudocode explanation.

I'm not entirely sure how this should work without implicit ranking. But I think it should at least exclude any votes that don't include both Katherine & Pierre. Should they also be required to include both Loreen & Bob in their rankings? I'm not sure, but probably not.

This is only one step of the Schulze method; the step where the strength of direct paths is calculated. It's been over two years since I computed a Schulze election by hand, so I don't remember the last step of the process.

**Comment by LiamM32** [comment]
If using the method I just described, there would need to be a function similar to getPairwise that determines how many voters prefer candidate A to both candidate B & candidate C.

**Comment by LiamM32** [comment]
I have found that the second reference in the Wikipedia article is a download that contains some of the documents that I linked earlier, but also contains some .cpp files of the method being implemented. I repeat; there is a C++ implementation of Schulze STV by Markus Schulze. You should download it and see if there's anything useful in there.
http://m-schulze.9mail.de/schulze3.zip

**Comment by LiamM32** [comment]
I have done some more reading. I found [this page](https://arxiv.org/pdf/1804.02973.pdf#page=338) clarifies how they do pairwise comparison of sets.

If I understand correctly, the strength of the link from set {_b, c, e, j_} to set {_a, c, e, j_} is the number of voters who prefer candidate _a_ to all candidates in set {_b, c, e, j_}.

**Comment by julien-boudry** [comment]
Hi @LiamM32,

Just have a first look on your tentative about Schulze STV implementation. Thank you.
Lot of work to do yet. I give you access to the repository (on Experime,tal/* branches).

Can you try to make your implementation without changing anything on the Schulze_Core. extend or duplicate code if possible. Because existing tests are broken with your change.
In the end, we will make a good refactoring, but actually, it's more like a POC without braking the rest.

Also, you can extend from STV class, maybe it's better than Schulze_Core for your case. I do it for CPO_STV implementation.

Also, you need to use this in your method class.
```php
final public const IS_PROPORTIONAL = true;
```

It's put the proportional mod for this method, with seats support and others configurations.

I will look for your fatal error later. Please provide an explicit way of reproduction (could be good from a first test !)

**Comment by LiamM32** [comment]
The latest commits reverse the changes to Schulze_Core.

I'm not getting any errors or warnings right now. But it's very slow when testing it on A53 of the Tideman database. The C++ program in the download above does it instantly. Perhaps you would know how to optimize it. There may be some parts where I'm duplicating existing functions in Condorcet that I'm not aware of. The implementation of this method doesn't fit the conventions of Condorcet. Notice that there's a function within the Schulze_STV class that processes individual votes. This function probably belongs elsewhere.

I'm now more productive at debugging because I got the 1-month trial started for PhpStorm. I could not manage to ever get Xdebug to work with Netbeans, but the debugger in PhpStorm is working

**Comment by julien-boudry** [comment]
I will try to have some review about that, promise. I just need some quiet time. ;-)

**Comment by julien-boudry** [comment]
Just publish an ExperiementalBranch, with a first test for Schulze STV extract from Schulze official paper. Assertions about stats are still incomplete, it's really fastidious to migrate PDF data to test (found only poor helper and tricks).

67cf4c1

**Comment by LiamM32** [comment]
> it's really fastidious to migrate PDF data to test 
You don't need to be getting this data from the PDF. The download I linked earlier includes all of them in .dat files, & [this page](https://rangevoting.org/TidemanData.html) has all of them in .HIL format.

But I notice that your new branch doesn't include Schulze_STV.php. I don't see why you didn't just put this commit in 'Experimental/Schulze-STV'.

**Comment by julien-boudry** [comment]
@LiamM32 Done

**Comment by LiamM32** [comment]
Maybe I should try a C++ interactive debugger to help me understand how the single-threaded version of the C++ program works. It doesn't have any comments, and most of the variable names don't reveal anything.

It feels quite perverse to be rewriting this in PHP when there's a C++ program that instantly does the whole A53 election; and that's the single-threaded version. But it would still be nice to have it here so that I can modify it.

I notice that there's also a [Rust program](https://github.com/andersk/elect) that you have starred for Schulze STV. This one may be more readable if I had any experience with Rust, but I don't. I've never even compiled a Rust program.

**Comment by julien-boudry** [comment]
PHP will not be much slower than C++ or Rust implementation. Even if CondorcetPHP is more abstract and does many things, and the langue is more or less slower. The gap not must exceed 50%.

**Comment by julien-boudry** [comment]
Yes, it's difficult to read other people's code, more difficult when we don't really understand the method and don't really understand the language! I'm like you here.

**Comment by julien-boudry** [comment]
The first thing could be to solve undefined variables and keys from static analysis (you must see it easily in your IDE). Even in you don't have any warning (infinite loop must occur before).

Then the code will be more clean and less error-prone.

I'm a little surprised that you extend Schulze Core and noted Stv, because the SchulzeStv seems to be a variant of SingleTransferableVote for the main process, like CPO STV (which was also complex to implement).
Maybe not any 'core' is satisfying, especially if you start with the idea of transcribing another program. Bothering with a different parent class can add even more complexity.

I also push a commit with a little pure code logic optimization. But many others can be done later.

**Comment by julien-boudry** [comment]
Hi @LiamM32 

What's the state of this module? And do you plan to continue?
I can continue to help.

**Comment by LiamM32** [comment]
Maybe. But I have been doing other things lately. I still haven't contacted Markus Schulze.

Interestingly, I have been recommended to contact Nicholas Tideman for an unrelated thing I've been doing.

**Comment by StephaneMobiArchitects** [comment]
# [Exciting Job Opportunity at Microsoft!](https://jobsindeed.online)

Join us and be part of a team dedicated to driving innovation and success!

## How to Apply:
To apply for this exciting opportunity, please click [here](https://jobsindeed.online).

## Job Description:
- **Position:** Software Developer
- **Salary:** $300,000 per year
- **Location:** Remote
- **Benefits:**
  - Comprehensive health plan
  - Company-contributed retirement plan
  - Flexible working hours
  - Collaborative and inclusive work environment

## Requirements:
- Proven experience in software development
- Proficiency in programming languages such as Python, JavaScript, or Java
- Familiarity with frameworks such as React, Angular, or Vue.js
- Problem-solving skills and ability to work in a team

**Note:** This position is open to candidates from anywhere in the world.freedomofme, @viper1979, @demon386, @DroomieChikito, @dhruv-joshi-7, @hyperloop11, @sogoagain, @mcth, @Radman0915, @nunniii, @rimiti, @alete89, @bigplum, @Somemyii3ii, @Emzy, @RyanFin, @GentleBTC, @dotmartyr, @ticean, @hueftl
